Paschen law as an introduction to low temperature plasmas

HollensteinCHH opened the pull request:
<!--
The present Paschen law project has some quite different aims, which will be explained in the following:

One aim is to introduce to PlasmaPy a first project concerning low temperature plasmas and discharge physics. Low temperature plasmas are a very large part of plasma physics and also strongly related to important industrial applications. Large interest in the plasma physics community exists and therefore projects concerning low temperature plasma should be developed in PlasmaPy.

The Paschen law is well known to plasma physicists and a good problem to start to introduce low temperature plasma topics into PlasmaPy and to show the different aspects this involves for PlasmaPy.

The famous Paschen law describes the breakdown voltage Vs in function of pd (gas pressure times electrode distance) for self-sustained small gap Townsend discharges.

The included (crude) Python program executes the following calculations:

Calculation of the breakdown voltage in function of the pressure/distance product for various gases and different electrode materials.

Short discussion of the so-called Paschen curve is included in the .py file.

Approximation to the Paschen curves are given and discussed.

The problem of the determination of the calculation of the parameters A and B (see .py file comments) is shown. To calculate these parameters access to material data base, cross sections data and even simulations such as the solutions to the Boltzmann equations is necessary.

It is this later part, which shows the upcoming problems to be solve or be fixed in relation to PlasmaPy in view of applications to low temperature plasmas.

How do we handle the different gases and even molecules or related issues in the frame of PlasmaPy?

How do we handle material properties within PlasmaPy, these data are necessary to calculate or simulate low temperature plasma physics?

How do we handle problems concerning cross sections and related data taken from well-known data bases or from calcualtions and simulations of low temperature plasmas found in literature.

The present simple problem for the determination of the breakdown voltage in self-sustained small gap Townsend discharge (Paschen law) is showing up intriguing problems for the implementation into PlasmaPy which have to be considered and be solved in future.d any new behavior.
* `bugfix`: Fixes a reported bug.
* `doc`: Documentation addition or improvement, like rewording an entire session or adding missing docs.
* `removal`: Feature deprecation and/or feature removal.
* `trivial`: A change which has no user facing effect or is tiny change.

A PR number is generated after you successfully submit a new PR, so the changelog
file can only be added after you open the PR.```

For more information, see:
https://github.com/PlasmaPy/PlasmaPy/blob/master/changelog/README.rst

--->

- [x] If adding new functionality, I have added tests and
      docstrings.

<!--
(Tests pop up at the bottom, in the checks box.)
-->

- [ ] I have fixed any newly failing tests.

<!--
(If you're unsure why they're failing, ask!)
-->

HollensteinCHH commented:
Please help me to intstall the pull request

rocco8773 commented:
@HollensteinCHH Welcome to the GitHub side of the conversation!  I hope the migration form emails to here wasn't too painful.  You even got the 1000th pull request.  That should be an achievement of some sort.  (on a side note...GitHub should have achievement/medals much like video games.  I wonder if that would increase contributions. 🤔)

It looks like you pull request (PR) came in just fine.  I see everything here.  There are two point I'd like to start with...

1. What is your intent with this PR?  We've done a little back and for with email, but I'm not sure if the purpose of this PR is to have a back and forth about the Paschen functionality to prototype it and work towards a final contribution in a separate PR, or are you hoping that this PR is the direct contribution to `plasmapy`? I guess the answer could be a little of both, but I'm just trying to gauge where you're at on it.
2. I feel this functionality would be better suited for `plasmapy.formulary` than `plasmapy.plasma`. @namurphy What is your thought on this?

HollensteinCHH commented:
Thanks for the comments. Paschen is in my intention a new part, section or however you name it in PlasmaPy which will deviate from the other, rather fusion related topics in PlasmaPy, towards  basic parts of plasma physics (by the way dispersion relation and Langmuir diagnostic are classic parts of basic plasma physics). The Paschen law is something all students in plasma physics know and therefore my contribution could attract students to correct and further develop my "first" Python program. As shown this simple problem can lead to quite interesting parts in plasma physics and to further developments. (in this case I think for field emission currents...  and so preparing the basis to more consequence fluid simulations of low temperature plasmas). Basic plasma physics has a long tradition in US (Prof Wong, Chen, Stenzel UCL) and also in Europe and Japan!
Simulation in plasma physics are complex just to say so and also many free software exists (UCLA particle in cell....)already and also in the fusion community a lot of codes around used. 
In adding Paschen in the Plasma folder I thought at the folder is intended to locate plasma physics  files (sorry about if I did wrong).
In order not to write page and page I will try to join tomorrow the meeting to discuss (hope it will work)

Christophe

namurphy commented:
Thank you, @HollensteinCHH!  It's very exciting to now have issue/pull request #1000.  I started to look through your pull request and it looks very interesting.  It's going to be great to have the Paschen law included in PlasmaPy.  

Of the checks that didn't pass, one appears to be related to a feature not available in the oldest version of NumPy we support.  The other issue is that we use a [pre-commit hook}(https://docs.plasmapy.org/en/stable/development/code_guide.html#pre-commit-hooks) to automatically format the code in a consistent way to help with readability.  Neither of these is a major problem and I'd be okay with waiting until later to address them.  We should also figure out where would be best to put the fundamental data needed for the Paschen law.

> Paschen is in my intention a new part, section or however you name it in PlasmaPy which will deviate from the other, rather fusion related topics in PlasmaPy, towards basic parts of plasma physics

I'm thinking that it would be good to have the essential functions in `plasmapy.formulary` for now, and then create an [educational Jupyter notebook](https://docs.plasmapy.org/en/stable/examples.html).  It would be worth thinking more about where to put it, since there will be more low temperature plasma functionality that we want to put in there later on.

> The Paschen law is something all students in plasma physics know and therefore my contribution could attract students to correct and further develop my "first" Python program. 

Most of my classes were on astronomy or plasma astrophysics, so I don't think we covered the Paschen law in there.  For the astronomers and heliophysicists among us, I think this contribution will be helpful as an educational resource too.

> In adding Paschen in the Plasma folder I thought at the folder is intended to locate plasma physics files (sorry about if I did wrong).

It's pretty routine to update pull requests after they were initially made, so this is no problem at all.  We'd be happy to talk with you more during [the meeting tomorrow](http://www.plasmapy.org/meetings/weekly/) if you can make it.  

Thank you again!

HollensteinCHH commented:
I added the Paschen1.py file to the plasmapy/formulary. However I was not able to delete the remaining file in the Plasma folder. 
As discussed yesterday I want to upload my  ipynb note book. Which folder?

Christoph

StanczakDominik commented:
If you place your notebook in `docs/examples`, it will automatically display among the others: https://plasmapy--1000.org.readthedocs.build/en/1000/examples.html :)

HollensteinCHH commented:
Dear all

I removed Paschen1.py from Plasma and added it to Formulary. I also added Paschen.ipynb to doc/notebook.

Dominik thanks for the patience and the great, great immediate help! 

Christoph

rocco8773 commented:
@HollensteinCHH The first thing we should do is rename `Paschen1.py` to `paschen.py` to remain consistent with [PEP8's package and module naming scheme](https://www.python.org/dev/peps/pep-0008/#package-and-module-names).  I have the documentation stuff ready to be merged into your branch and I'll do that once the file is renamed.  After that we can get into doing reviews directly on the code.

rocco8773 commented:
Additionally, ...

1. The functionality contained in `paschen.py` should focus on calculating the breakdown voltage, and not any plotting.  The plotting can be left in the notebook though.  We've been avoiding the practice of providing plotters since most users will plot their own figures.
2. I think the Paschen functionality can be reduced to one function that looks something like...
    ```python
    def breakdown_voltage(distance, pressure, A, B, townsend_gamma):
        ...
        return vb
    ```
    where `distance` and `pressure` could be scalars or a 1D `numpy` array, `A` and `B` are the material constants, and `townsend_gamma` is the second Townsend coefficient.  I used `townsend_gamma` instead of `gamma` since the formulary already uses `gamma` for the adiabatic constant.
3. I'd really like to incorporate the material constants into the `Particle` class to simplify the signature of `breakdown_voltage()`, but I think that should be left of a follow up PR.

HollensteinCHH commented:
I agree completely with this changes. I will make the necessary  as soon as possible

HollensteinCHH commented:
I change the name from Paschen1 to Paschen!

rocco8773 commented:
> I change the name from Paschen1 to Paschen!

Can you make it `paschen.py` please?  The notebook can still be `Paschen.ipynb`.  Python's PEP8 standards reserve CamelCase for classes, so having the module name `Paschen` may lead to miss interpretation by users.

HollensteinCHH commented:
Sorry,I changed! Thanks

rocco8773 commented:
Awesome!  Now, the way `paschen.py` is currently structured is causing issues will building the documentation.  So, I'm not going to push the documentation changes until you have gone through the file consolidating the code and eliminating the plotters as per my previous [comment](https://github.com/PlasmaPy/PlasmaPy/pull/1000#issuecomment-784724591).  Also, once you do that then I can start giving more focused review comments.  Thanks!

HollensteinCHH commented:
Thanks for the help. I wanted to start to change the paschen.py according to your suggestions.

I still agree, that in formulary no plotting! This is ok!

However

""
I think the Paschen functionality can be reduced to one function that looks something like...

def breakdown_voltage(distance, pressure, A, B, townsend_gamma):

    ...

    return vb

""
I think such a functionality does not make much sense. Why? The user has already to know the different parameters A, B and “gamma”. To find these parameters is the main work in calculating the breakdown voltage, once you have them it is easy to calculate and then you don’t need PlasmaPy to do it! (the same remark is also valuable for saha.py, where g_k's and Ejk are input parameters, the following calculation is trivial, the real bottleneck are the g's respectively the partition functions hard to determine, it is here PlasmaPy must assist the user.) 

The idea proposing to PlasmaPy Paschen was that PlasmaPy opens towards other plasmas then just hydrogen plasmas. Ok, they are important in astro and fusion but there are many other plasmas also playing important roles in education and applications. For this reason, PlasmaPy must have a sort of “data base” for other plasma relevant parameters for different elements (gas, electrode materials, excitation coefficients, cross sections…). Often in calculating or simulating plasmas such info is missing or hard to find, here PlasmaPy could close an important gap and collaboration with OtherPy’s could or should be find.

For the formulary I propose

def breakdown voltage (distance,pressure, gas, electrode_material )

Such a functionality could be really more helpful for the user (mostly students) or should we omit it in the formulary and only make just a Paschen notebook?

rocco8773 commented:
@HollensteinCHH You're right that finding the constants is a significant part in calculating the breakdown voltage.  PlasmaPy should have a database of values, but we will never have an exhaustive database of all the possible values.  For this reason, the `breakdown_voltage()` function has to still have input parameters for manually passing in these constants.

Now, how to properly implement this is up for discussion? My first stab at would look something like...

```python
def get_paschen_constants(gas, electrode):
    # this will search the database for the constants contained in PlasmaPy
    ...
    return dict_of_constants

def breakdown_voltage(distance, pressure, A, B, townsend_gamma):
    # calculate the breakdown voltage
    ...
    return vb
```

Then you would use the functionality like...

```python
>>> const = get_paschen_constants(gas="He", electrode="Al")
>>> const
{"A": 2.3, "B": 26, "townsend_gamma": 0.021}
>>> vb = breakdown_voltage(distance, pressure, **const)
>>>
>>> # you can still do manual entries
>>> vb = breakdown_voltage(distance, pressure, A=2.01, B=26.5, townsend_gamma=0.03)
```

---

> I think such a functionality does not make much sense. Why? The user has already to know the different parameters A, B and “gamma”. To find these parameters is the main work in calculating the breakdown voltage, once you have them it is easy to calculate and then you don’t need PlasmaPy to do it!

I mostly disagree with this statement because it only comes from a perspective of offering functionality.  A goal of PlasmaPy is to be an educational resource through it's robust documentation.  Even if some functionality only implements a "trivial" expression, it is still useful to have if the documentation is thorough enough to explain the physics behind said functionality.

namurphy commented:
> How do we realise a internal data base with PlasmaPy?

This is a good question!  How we provide fundamental data is a long-term problem that we still need to figure out.  This pull request does not include much data, so it is fine to keep it in `paschen.py` for this pull request.

> Presently I am busy with some consulting and other works, but as soon as possible I will change the formulary in the sense we discussed, but we need to solve also the problems around a data base within PlasmaPy for future projects. 

Thank you!

namurphy commented:
The next step for this pull request is to format the functionality into the style that we use for PlasmaPy.  It might be helpful to look at an example file, [`atomic.py`](https://github.com/PlasmaPy/PlasmaPy/blob/master/plasmapy/particles/atomic.py), to see how the documentation strings at the beginning of each function are structured.  We use the [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard) style for docstrings, and the docstrings are formatted using [reStructuredText](https://docutils.sourceforge.io/rst.html).  

A few suggestions would be to:
 - Put a one line description on the first line of the file, like: ``"""Functionality related to the Paschen law."""``
 - Move the plotting functionality out of `paschen.py` and into the example notebook
 - Format function names like to be like: `minimum_breakdown_voltage` (lower case with words separated by underscores)
 - Move the documentation strings to start the line after the `def` statement, with lines indented by four spaces.  Lines in docstrings should be limited to roughly 72–80 characters.
 - Start each documentation string with a one or two line statement of what each function does.  
 - Put the extended description in a `Notes` section near the end of the documentation string, like for [`standard_atomic_weight`](https://github.com/PlasmaPy/PlasmaPy/blob/6a161f09530ee3d6870ea00a5c95f5ef278e12fb/plasmapy/particles/atomic.py#L169)

Thank you again!

HollensteinCHH commented:
I agree with the comments and I will try to make the necessary changes.

HollensteinCHH commented:
I did a first change to paschen.py, according to your comments. I will now start to make it compatible to PlasmaPy

HollensteinCHH commented:
Here. an other iteration!

HollensteinCHH commented:
I think I need some help, please!

namurphy reviewed (changes requested):
Thank you again for working on this!  I used GitHub's "suggest" feature to propose a few changes, which you would need to accept before the change is made.  Most of the suggested changes are related to syntax.  The [coding style guide](https://www.python.org/dev/peps/pep-0008/) we use has lines inside functions being indented by increments of four spaces.  Python is pretty strict about indentation...lines that are in the same block should be indented at the same level.  

Some of the next steps are:
 - In `plasmapy/formulary/__init__.py`, add a line that says `from .paschen import *`, which will allow the functions in `paschen.py` to be imported from `plasmapy.formulary`.
 - At the top of `plasmapy/formulary/paschen.py` (like in [this file](https://github.com/PlasmaPy/PlasmaPy/blob/5eb8450cc8a46fd5ba5ef86fac28115271978c70/plasmapy/analysis/swept_langmuir/floating_potential.py#L1))
   - Add a short line in triple quotes that describes the purpose of the file
   - Below that, add a line like `__all__ = [   ]`.  Inside the brackets, put in strings with the name of each of the functions.  This is to tell the documentation generator which functions to include in the documentation.
 - Create a file: `plasmapy/formulary/tests/test_paschen.py`.  This is the file where we'll need to add tests.  Our [documentation on tests](https://docs.plasmapy.org/en/stable/development/testing_guide.html#writing-tests) has some more information about how to write tests.

HollensteinCHH commented:
Hello

Thanks for the great help! I will make the necessary corrections. I start getting  the feeling on the style demanded by PlasmaPy. I read many manuals and tutorials (ufff) putting them them in practice is an other story. 
I might get in contact with a former PhD student which could be interested to do some work and also he would have connections to students. 
I will also continue to prepare some future works in thermal plasma composition and thermoionc emission…

Thanks


Christoph 


> On 17 Mar 2021, at 22:34, Nick Murphy ***@***.***> wrote:
> 
> 
> @namurphy requested changes on this pull request.
> 
> Thank you again for working on this! I used GitHub's "suggest" feature to propose a few changes, which you would need to accept before the change is made. Most of the suggested changes are related to syntax. The coding style guide <https://www.python.org/dev/peps/pep-0008/> we use has lines inside functions being indented by increments of four spaces. Python is pretty strict about indentation...lines that are in the same block should be indented at the same level.
> 
> Some of the next steps are:
> 
> In plasmapy/formulary/__init__.py, add a line that says from .paschen import *, which will allow the functions in paschen.py to be imported from plasmapy.formulary.
> At the top of plasmapy/formulary/paschen.py (like in this file <https://github.com/PlasmaPy/PlasmaPy/blob/5eb8450cc8a46fd5ba5ef86fac28115271978c70/plasmapy/analysis/swept_langmuir/floating_potential.py#L1>)
> Add a short line in triple quotes that describes the purpose of the file
> Below that, add a line like __all__ = [ ]. Inside the brackets, put in strings with the name of each of the functions. This is to tell the documentation generator which functions to include in the documentation.
> Create a file: plasmapy/formulary/tests/test_paschen.py. This is the file where we'll need to add tests. Our documentation on tests <https://docs.plasmapy.org/en/stable/development/testing_guide.html#writing-tests> has some more information about how to write tests.
> In plasmapy/formulary/paschen.py <https://github.com/PlasmaPy/PlasmaPy/pull/1000#discussion_r596116379>:
> 
> > +
> +  r"""
> + 
> +      Function to get the constants A and B and the second Townsend coefficient to calculate the Paschen breakdown voltage
> +   
> +                     
> +
> ⬇️ Suggested change
> -
> -  r"""
> - 
> -      Function to get the constants A and B and the second Townsend coefficient to calculate the Paschen breakdown voltage
> -   
> -                     
> -
> +    r"""
> +    Return the constants A and B and the second Townsend coefficient to calculate
> +    the Paschen breakdown voltage.                   
> [docstyle] Lines in docstrings for regular text should be indented four spaces. We also use a convention to start the first line of a docstring with a verb, typically "Return" or "Calculate".
> 
> In plasmapy/formulary/paschen.py <https://github.com/PlasmaPy/PlasmaPy/pull/1000#discussion_r596142262>:
> 
> > +      Parameters
> +      ----------
> +
> +      gas :        'str'
> +      electrode :  'str'
> +        String representing the gas and electrode material
> ⬇️ Suggested change
> -      Parameters
> -      ----------
> -
> -      gas :        'str'
> -      electrode :  'str'
> -        String representing the gas and electrode material
> +    Parameters
> +    ----------
> +    gas : `str`
> +        The gas for which to find the coefficients.
> +        
> +    electrode : `str`
> +        The electrode material.
> This suggestion puts the syntax into the form that is expected by our documentation generator, e.g.:
> 
>     parameter_name : `type`
>         Short description of parameter, indented by another four spaces.
> The single back ticks ` tell our documentation generator to create a link to the object (e.g., `str` will link to the Python documentation for a string object).
> 
> In plasmapy/formulary/paschen.py <https://github.com/PlasmaPy/PlasmaPy/pull/1000#discussion_r596281613>:
> 
> > +      Examples
> +      --------
> +
> +      c=def get_paschen_constants ("Ar","Ni):
> +      c={'A': 11, 'B': 135, 'gam': 0.058} 
> +
> +      c=def get_paschen_constants ("Ar","zz"):
> +      c={'A': 11, 'B': 135, 'gam': 0.01}
> +      If electrode material is not found a default value of 0.01 is taken
> +
> +      c=def get_paschen_constants ("Zz","Ni"):
> +      c=None
> +      If gas is not found, c is set to None 
> ⬇️ Suggested change
> -      Examples
> -      --------
> -
> -      c=def get_paschen_constants ("Ar","Ni):
> -      c={'A': 11, 'B': 135, 'gam': 0.058} 
> -
> -      c=def get_paschen_constants ("Ar","zz"):
> -      c={'A': 11, 'B': 135, 'gam': 0.01}
> -      If electrode material is not found a default value of 0.01 is taken
> -
> -      c=def get_paschen_constants ("Zz","Ni"):
> -      c=None
> -      If gas is not found, c is set to None 
> +    Examples
> +    --------
> +    >>> get_paschen_constants("Ar", "Ni)
> +    {'A': 11, 'B': 135, 'gam': 0.058} 
> +
> +    If the electrode material is not found, then a default value of 0.01 is taken.
> +
> +    >>> get_paschen_constants("Ar", "zz")
> +    {'A': 11, 'B': 135, 'gam': 0.01}
> +
> +    If ``gas`` is not found, ``c`` is set to `None`.
> +    
> +    >>> get_paschen_constants("Zz", "Ni")
> +    None
> In examples, the Python code that a user would type follows >>> . The output gets put on the line after that. The testing framework we use checks that the output matches exactly the output expected from running the line above. The double back ticks (like around ``gas``) indicate that the text inside should be formatted like code, except without doing any linking.
> 
> In plasmapy/formulary/paschen.py <https://github.com/PlasmaPy/PlasmaPy/pull/1000#discussion_r596295167>:
> 
> > +      Return
> +      ------
> +
> +      dictionary containing the constants A, B and townsend_gamma for calculation of the breakdwn voltage
> ⬇️ Suggested change
> -      Return
> -      ------
> -
> -      dictionary containing the constants A, B and townsend_gamma for calculation of the breakdwn voltage
> +    Returns
> +    -------
> +    Dictionary containing the constants ``A``, ``B`` and ``townsend_gamma`` for calculation
> +    of the breakdown voltage.
> A few more minor syntax changes to put the docstring into the form expected by our documentation generator. The slight change in section name is for consistency with elsewhere in the docs.
> 
> In plasmapy/formulary/paschen.py <https://github.com/PlasmaPy/PlasmaPy/pull/1000#discussion_r596298686>:
> 
> > +      References
> +      ---------
> +      Paschen_constants contains the coefficents A and B  for the estimation of the
> +      First Townsend Ionization Coefficent
> +       (exponential fit to the First Townsend Ionization coefficient)
> +       as adapted from
> +       E.Nasser, Fundamentals of Gaseous Ionization and Plasma Electronics,
> +       Wiley-Interscience, New York 1971
> +    
> +      format: paschen_constants dir {"gas":[A,B]}
> +      units: A in [Ionisation/(Pa m)] and B in [V/(Pa m)]
> +
> +
> +      Townsend_gamma is the Second Townsend Ionization coefficient as given by
> +        A.Beroual and I. Fonfana, Discharge in Long Air Gap Modeling and Application
> +        IOP Publishing Ltd 2016
> +        ISBN 978-0-7503-1236-3 (ebook)
> +        ISBN 978-0-7503-1237-0 (print)
> +
> +
> ⬇️ Suggested change
> -      References
> -      ---------
> -      Paschen_constants contains the coefficents A and B  for the estimation of the
> -      First Townsend Ionization Coefficent
> -       (exponential fit to the First Townsend Ionization coefficient)
> -       as adapted from
> -       E.Nasser, Fundamentals of Gaseous Ionization and Plasma Electronics,
> -       Wiley-Interscience, New York 1971
> -    
> -      format: paschen_constants dir {"gas":[A,B]}
> -      units: A in [Ionisation/(Pa m)] and B in [V/(Pa m)]
> -
> -
> -      Townsend_gamma is the Second Townsend Ionization coefficient as given by
> -        A.Beroual and I. Fonfana, Discharge in Long Air Gap Modeling and Application
> -        IOP Publishing Ltd 2016
> -        ISBN 978-0-7503-1236-3 (ebook)
> -        ISBN 978-0-7503-1237-0 (print)
> -
> -
> +    References
> +    ---------
> +    Paschen_constants contains the coefficents A and B  for the estimation of the
> +    First Townsend Ionization Coefficent
> +    (exponential fit to the First Townsend Ionization coefficient)
> +    as adapted from
> +    E.Nasser, Fundamentals of Gaseous Ionization and Plasma Electronics,
> +    Wiley-Interscience, New York 1971
> +    
> +    format: paschen_constants dir {"gas":[A,B]}
> +    units: A in [Ionisation/(Pa m)] and B in [V/(Pa m)]
> +
> +    `townsend_gamma` is the Second Townsend Ionization coefficient as given by
> +    A.Beroual and I. Fonfana, Discharge in Long Air Gap Modeling and Application
> +    IOP Publishing Ltd 2016
> +    ISBN 978-0-7503-1236-3 (ebook)
> +    ISBN 978-0-7503-1237-0 (print)
> Here's a minor syntax change to have the indent be four spaces. I can never remember what the formatting style we use for references is, but that's a minor change that can wait until later.
> 
> In plasmapy/formulary/paschen.py <https://github.com/PlasmaPy/PlasmaPy/pull/1000#discussion_r596300247>:
> 
> > +  gases=["Air","N2","H2","He","Ne","Ar","Kr","Xe"]   
> +
> +  paschen_constants={"Air":[11,274],
> ⬇️ Suggested change
> -  gases=["Air","N2","H2","He","Ne","Ar","Kr","Xe"]   
> -
> -  paschen_constants={"Air":[11,274],
> +    gases=["Air","N2","H2","He","Ne","Ar","Kr","Xe"]   
> +
> +    paschen_constants={"Air":[11,274],
> Minor syntax suggestion.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/PlasmaPy/PlasmaPy/pull/1000#pullrequestreview-614446372>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ASV5UTZKBZQTYHQKX2CTYW3TEEN6TANCNFSM4XGOPTVA>.
>

HollensteinCHH commented:
I have difficulties to find the errors indicated by all the tests. All the errors are coming from the documentation part and not from the main program. The program (also the notebook) is running on Colab. Yes, I used Colab (as you intended to use for PlasmaHack) since I can not install all the different programs, since if something fails I will get IT problems. I have no IT help and equipment as many.
 By the way I chose the extremely simple topic of Paschen just to learn Python and get used to Git and PlasmaPy. 
Is there somebody (students...)  which might me give some help, the code is not really, really complex?
Thanks!

StanczakDominik commented:
Well, colab is certainly an unorthodox method of development but hey, if it works, it works! I just checked and according to https://stackoverflow.com/questions/56577003/git-pull-on-google-colab you can in fact interact with git from it, so it should be okay.

By the way, the suggestions Nick made above, if you read them from the GitHub interface, such as https://github.com/PlasmaPy/PlasmaPy/pull/1000/files?file-filters%5B%5D=.py&file-filters%5B%5D=.rst#r596300247 here, you can then apply them to your code right there. You don't have to reapply them manually.

I'll look through these and try to fix the test failures. I can either turn any changes I make into that kind of code suggestions, or I can push them to your branch; then you can use `git pull` to integrate them back into your local (or colab) repository. just say which one you prefer, @HollensteinCHH :)

StanczakDominik commented:
@namurphy @rocco8773 a side note, because there's no way I'm not diving on this opportunity: if we were to run pre-commit.ci, @HollensteinCHH wouldn't have to make the syntax fixes himself or install pre-commit at all, just git pull the changes (which is already the case for suggestions), and we would see the cleaned up version in review ;)

StanczakDominik commented:
Okay, I can't make a clean pull request to your branch, since I made a merge from master to this branch along the way and it's going to show way too many commits for this to be easily readable.

Instead, I'm making a push to this branch. You will have to pull it to your local directory (that should be as simple as `%cd /location/of/plasmapy/repository; !git pull` from a cell in Colab. I think.

If that doesn't work, give me a heads up and I'll help you through it.

---

I've made a bunch of changes:

* b63bdaf applies our automated code style fixers to your changes. This makes the code cleaner and easier for me to look through.
* 32b5eec changes how the functions are called in their examples. You forgot to remove `def` from there, so the syntax in the examples was indeed invalid.
* 3bedcf5 removes print statements. We generally don't want functions to emit external noise except in longer simulations, etc, where you actually want to notify the user that yes, something is happening that's not just a `while True` loop :) I understand they're handy for debugging, but they're going to have to go some time before merging, anyway.
* 8a3fe99 adjusts the section name to the usual Parameters (plural).
* e976633 is a merge commit; it merges all the changes that have been happening on the master branch into yours, so that your current state of work is up to date with the rest of the repository.
* 65d77c8, as described in the commit message up top, turns long comments in notebook code blocks into Markdown cells (they're easier on the eyes!), and replaces `nonposy` with `nonpositive` in Matplotlib plotting calls. Matplotlib is deprecating nonposy with a warning since Matplotlib 3.3,
so it'd be noisy to keep it that way.
* and finally, d3ed25e changes the math formatting in the notebook a little. I'm not entirely sure why sphinx was complaining about it, but it's working now. :3

StanczakDominik commented on plasmapy/formulary/paschen.py:
```diff
@@ -0,0 +1,279 @@
+            #       If not a default townsend_gamma value of 0.01 is taken
+
+            if gn is None:
+                gn = 0.01
```
I have literally no expertise with this area. Is 0.01 a sensible value "in general"? Perhaps, rather than assume it, we could raise a `KeyError: no data for the given gas/electrode`, or at the very least throw a warning and say that we're assuming the default. I think I'd prefer the former approach.

HollensteinCHH commented:
Thanks Dominik!
I will go through all this, I want to learn out of my errors and hope to be more helpful in future.

Yes I know Colab is a unconventional way, but many students will try going such a way, in particular in this hard pandemic times. If we can get them doing some useful work, would be nice for them and also us.

StanczakDominik commented on plasmapy/formulary/paschen.py:
```diff
@@ -0,0 +1,279 @@
+            "W": None,
+            "Fe": 0.015,
+        },
+        "Ne": {
+            "Al": 0.053,
+            "Cu": 0.02,
+            "Ni": 0.031,
+            "Pt": 0.023,
+            "C": None,
+            "W": 0.045,
+            "Fe": 0.022,
+        },
+        "Ar": {
+            "Al": 0.12,
+            "Cu": 0.058,
+            "Ni": 0.058,
+            "Pt": 0.058,
+            "C": None,
+            "W": None,
+            "Fe": 0.058,
+        },
+        "Kr": {
+            "Al": None,
+            "Cu": None,
+            "Ni": None,
+            "Pt": None,
+            "C": None,
+            "W": None,
+            "Fe": None,
+        },
+        "Xe": {
+            "Al": None,
+            "Cu": None,
+            "Ni": None,
+            "Pt": None,
+            "C": None,
+            "W": None,
+            "Fe": None,
+        },
+    }
```
If we do go with exceptions or warnings, then we could drop all the `None`s and replace the latter part with a `try/except KeyError` block. Simple to code that up, so I could do it for you, but I'd rather get agreement on how best to proceed here first.

StanczakDominik commented on plasmapy/formulary/paschen.py:
```diff
@@ -0,0 +1,279 @@
+
+    """
+
+    #  Calculate breakdown voltage according to the Paschen law
+    x = list(distance * p for p in pressure)
+    g = np.log(1 + (1 / gam))
+    vb = list((pd, B * pd / (np.log((A * pd) / g))) for pd in x)
+    return vb
```
It might be easier and faster to require a numpy array rather than the list, and return a 2-tuple of numpy arrays rather than a list of 2-tuples.

StanczakDominik commented on plasmapy/formulary/paschen.py:
```diff
@@ -0,0 +1,279 @@
+    """
+    # Calculate vmin and pdmin
+    g = np.log(1 + (1 / gam))
+    vmin = 2.718 * (B / A) * g
+    pdmin = 2.718 * (g / A)
```
I would definitely appreciate some comments here on the source for the magic 2.718 number :smile:

StanczakDominik replied:
since this is `e`, 
```suggestion
    vmin = np.e * (B / A) * g
    pdmin = np.e * (g / A)
```

StanczakDominik reviewed (changes requested):
Having made the formatting/CI fixes, here's some design comments!

HollensteinCHH commented:
Nothing magic in physics! I wanted to change this since a long time but this stands for e!

StanczakDominik commented:
> Nothing magic in physics! I wanted to change this since a long time but this stands for e!

............................. okay, I guess I'm officially still not fully recovered from yesterday's neoclassical all-nighter :laughing:

StanczakDominik commented:
In that case, I added a suggestion you can try applying from the Files Changed tab - should be good practice ;)

HollensteinCHH commented:
I will try, hope to succeed!
Nice evening 

> On 27 Mar 2021, at 19:01, Dominik Stańczak ***@***.***> wrote:
> 
> 
> In that case, I added a suggestion you can try applying from the Files Changed tab - should be good practice ;)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/PlasmaPy/PlasmaPy/pull/1000#issuecomment-808770931>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ASV5UTZHRM46GYVLONFZQH3TFYMRPANCNFSM4XGOPTVA>.
>

HollensteinCHH reviewed (commented):
Sorry, I just remarked that all the following remarks have been pending due to a not submitted blank comment in the long list which I didn't check lack to experience with Git.(I have still to learn a lot, I will try...). I might have done some erroneous manipulation. Sorry about this and the delay due to that. 

Hope now to continue again with the Paschen formulary and the notebook and contribute finally too PlasmaPy.  
Since I was stuck I started with a notebook calculating plasma composition with Saha (I have already a part with H2 dissociation and following H ionisation)
I made publicity for PlasmaPy on some universities and might get some interested people (students and supervisors).

namurphy commented:
Thank you, @HollensteinCHH!  I'm excited about this pull request so I'll try to respond in more detail in the next few days on this pull request and in your email.  If you'd like we could set up a time to meet in the next week or two over Zoom (especially since I tend to be slow to respond to email).

> Sorry, I just remarked that all the following remarks have been pending due to a not submitted blank comment in the long list which I didn't check lack to experience with Git.(I have still to learn a lot, I will try...).

I do the exact same that at a rate of ∼0.5 month⁻¹, so I definitely still have a lot to learn too...

> I made publicity for PlasmaPy on some universities and might get some interested people (students and supervisors).

This is great news, thank you!

HollensteinCHH commented:
Thank you for the mail and your understanding. Yes would be nice to set up a Zoom, this is faster than pages and pages of mail. I don.t have much obligations, so please feel free to propose date and hour. 

Best regards from Switzerland 

Christoph 



> On 5 May 2021, at 15:31, Nick Murphy ***@***.***> wrote:
> 
> 
> Thank you, @HollensteinCHH <https://github.com/HollensteinCHH>! I'm excited about this pull request so I'll try to respond in more detail in the next few days on this pull request and in your email. If you'd like we could set up a time to meet in the next week or two over Zoom (especially since I tend to be slow to respond to email).
> 
> Sorry, I just remarked that all the following remarks have been pending due to a not submitted blank comment in the long list which I didn't check lack to experience with Git.(I have still to learn a lot, I will try...).
> 
> I do the exact same that at a rate of ∼0.5 month⁻¹, so I definitely still have a lot to learn too...
> 
> I made publicity for PlasmaPy on some universities and might get some interested people (students and supervisors).
> 
> This is great news, thank you!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/PlasmaPy/PlasmaPy/pull/1000#issuecomment-832689147>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ASV5UT23KXOOSZ5IDXZK5ZTTMFCCRANCNFSM4XGOPTVA>.
>

G-D-Zhou commented:
For me who has a strong interest in low-temperature plasma applications, this pull request (#1000) and further going topics are enough to make me too excited to sleep. Tomorrow, I will carefully read the example of the Paschen law you submitted and try to rewrite calculations by myself for practice.

G-D-Zhou commented:
I followed what I said before and read part of your code. I made certain changes to it to make it more readable. I strongly recommend creating a special python file for the secondary electron coefficients for different materials, A, and B, which may be beneficial to the realization of your subsequent topics. In this regard, I made a preliminary attempt. Unfortunately, I do not know how to use Git to upload my code to plasmapy. Therefore, it cannot be shown to you directly. Moreover, I don’t know whether adding other links to this site is in compliance with the rules. 
Anyway, thank you very much for introducing Paschen's law. Although the law is, as you said, very common for students of low-temperature plasma physics. But for non-major students, for example, me, the help is particularly great. It reasonably explained the problems I encountered in the experiment and promoted me to continue to use my spare time to learn low-temperature plasma physics simulation. Finally, I look forward to your development of a two-dimensional DC plasma simulation code based plasmapy. I will continue to follow your work.

HollensteinCHH commented:
Thank you (G-D-Zhou) very much for your kind remarks concerning the Paschen contribution to Plasmapy. Indeed I am not “the”expert in Python (by the way also not of GitHub)  but I tried to write a more or less acceptable code as a starting point. So, any changes and helps are welcome, which I will all consider. I hope to contribute mainly with my knowledge in plasma physics in particular also of the low temperature plasmas and the various industrial plasmas and I am thankful for people helping me to bring this in a form for PlasmaPy and also as a help to students. In fact I also started to interest here some labs for interesting works both for PlasmaPy and plasma physics (but also IT ) education. Maybe I will have in future some help here around me!  
Beside the Paschen contribution I am presently working at a contribution using the Saha equation and to calculate the plasma composition of various atomic and molecular thermal plasmas (at present hydrogen dissociation and ionization is starting to run, it take more time than expected!). I hope to develop the topic of non-thermal plasmas in PlasmaPy further (I hope), there are so many topics we could include into PlasmaPy.
You are right, non-thermal plasmas use many “outside”data (such as the coefficient A and B, but there are many, many more such as cross sections, atomic and molecular properites ….) which have to be included somewhere in PlasmaPy. I am in contact with Nick, Eric and Dominik for that, that is a topic for “Python”.

Would be glad to stay in contact!

Best regards 

Christoph 
Thank you (G.Zhou) very much for your kind remarks concerning the Paschen contribution to Plasmapy. Indeed I am not “the”expert in Python (by the way also not of GitHub)  but I tried to write a more or less acceptable code as a starting point. So, any changes and helps are welcome, which I will all consider. I hope to contribute mainly with my knowledge in plasma physics in particular also of the low temperature plasmas and the various industrial plasmas and I am thankful for people helping me to bring this in a form for PlasmaPy and also as a help to students. In fact I also started to interest here some labs for interesting works both for PlasmaPy and plasma physics (but also IT ) education. Maybe I will have in future some help here around me!  
Beside the Paschen contribution I am presently working at a contribution using the Saha equation and to calculate the plasma composition of various atomic and molecular thermal plasmas (at present hydrogen dissociation and ionization is starting to run, it take more time than expected!). I hope to develop the topic of non-thermal plasmas in PlasmaPy further (I hope), there are so many topics we could include into PlasmaPy.
You are right, non-thermal plasmas use many “outside”data (such as the coefficient A and B, but there are many, many more such as cross sections, atomic and molecular properites ….) which have to be included somewhere in PlasmaPy. I am in contact with Nick, Eric and Dominik for that, that is a topic for “Python”.

Would be glad to stay in contact!

Best regards 

Christoph

G-D-Zhou commented:
> Thank you (G-D-Zhou) very much for your kind remarks concerning the Paschen contribution to Plasmapy. Indeed I am not “the”expert in Python (by the way also not of GitHub) but I tried to write a more or less acceptable code as a starting point. So, any changes and helps are welcome, which I will all consider. I hope to contribute mainly with my knowledge in plasma physics in particular also of the low temperature plasmas and the various industrial plasmas and I am thankful for people helping me to bring this in a form for PlasmaPy and also as a help to students. In fact I also started to interest here some labs for interesting works both for PlasmaPy and plasma physics (but also IT ) education. Maybe I will have in future some help here around me!
> Beside the Paschen contribution I am presently working at a contribution using the Saha equation and to calculate the plasma composition of various atomic and molecular thermal plasmas (at present hydrogen dissociation and ionization is starting to run, it take more time than expected!). I hope to develop the topic of non-thermal plasmas in PlasmaPy further (I hope), there are so many topics we could include into PlasmaPy.
> You are right, non-thermal plasmas use many “outside”data (such as the coefficient A and B, but there are many, many more such as cross sections, atomic and molecular properites ….) which have to be included somewhere in PlasmaPy. I am in contact with Nick, Eric and Dominik for that, that is a topic for “Python”.
> 
> Would be glad to stay in contact!
> 
> Best regards
> 
> Christoph
> Thank you (G.Zhou) very much for your kind remarks concerning the Paschen contribution to Plasmapy. Indeed I am not “the”expert in Python (by the way also not of GitHub) but I tried to write a more or less acceptable code as a starting point. So, any changes and helps are welcome, which I will all consider. I hope to contribute mainly with my knowledge in plasma physics in particular also of the low temperature plasmas and the various industrial plasmas and I am thankful for people helping me to bring this in a form for PlasmaPy and also as a help to students. In fact I also started to interest here some labs for interesting works both for PlasmaPy and plasma physics (but also IT ) education. Maybe I will have in future some help here around me!
> Beside the Paschen contribution I am presently working at a contribution using the Saha equation and to calculate the plasma composition of various atomic and molecular thermal plasmas (at present hydrogen dissociation and ionization is starting to run, it take more time than expected!). I hope to develop the topic of non-thermal plasmas in PlasmaPy further (I hope), there are so many topics we could include into PlasmaPy.
> You are right, non-thermal plasmas use many “outside”data (such as the coefficient A and B, but there are many, many more such as cross sections, atomic and molecular properites ….) which have to be included somewhere in PlasmaPy. I am in contact with Nick, Eric and Dominik for that, that is a topic for “Python”.
> 
> Would be glad to stay in contact!
> 
> Best regards
> 
> Christoph

Hello, Christoph.
I am very happy to receive your reply. I will write the code modified slightly in the next comment, which may cause the comment to be very long. Therefore, after you read and reply, I will delete the code. Hope these can contribute to plasmapy.
Looking forward to your next work!
Would be glad to stay in contact!
Best regards
G. Zhou

G-D-Zhou commented:
constant.py
`
class Constant_Paschen():
    def constant_A_B_Paschen():
        # Paschen....[A,B] A [Ionisation/(Pa m)] B [V/(Pa m)]
        constant_A_B_Paschen = {
            "Air": [11, 274],
            "N2": [9.0, 257],
            "H2": [3.8, 104],
            "He": [2.3, 26],
            "Ne": [3.0, 75],
            "Ar": [11, 135],
            "Kr": [13, 180],
            "Xe": [20, 263],
        }
        return constant_A_B_Paschen  
    def gamma():
        gamma = {
        "Air": {
            "Al": 0.035,
            "Cu": 0.025,
            "Ni": 0.036,
            "Pt": 0.017,
            "C": None,
            "W": None,
            "Fe": 0.02,
        },
        "N2": {
            "Al": 0.1,
            "Cu": 0.066,
            "Ni": 0.077,
            "Pt": 0.59,
            "C": None,
            "W": None,
            "Fe": 0.059,
        },
        "H2": {
            "Al": 0.095,
            "Cu": 0.05,
            "Ni": 0.053,
            "Pt": 0.02,
            "C": 0.014,
            "W": None,
            "Fe": 0.061,
        },
        "He": {
            "Al": 0.021,
            "Cu": None,
            "Ni": 0.015,
            "Pt": 0.01,
            "C": None,
            "W": None,
            "Fe": 0.015,
        },
        "Ne": {
            "Al": 0.053,
            "Cu": 0.02,
            "Ni": 0.031,
            "Pt": 0.023,
            "C": None,
            "W": 0.045,
            "Fe": 0.022,
        },
        "Ar": {
            "Al": 0.12,
            "Cu": 0.058,
            "Ni": 0.058,
            "Pt": 0.058,
            "C": None,
            "W": None,
            "Fe": 0.058,
        },
        "Kr": {
            "Al": None,
            "Cu": None,
            "Ni": None,
            "Pt": None,
            "C": None,
            "W": None,
            "Fe": None,
        },
        "Xe": {
            "Al": None,
            "Cu": None,
            "Ni": None,
            "Pt": None,
            "C": None,
            "W": None,
            "Fe": None,
        },
    }
        return gamma
`

G-D-Zhou commented:
```python
markdown can be works?
```

G-D-Zhou commented:
constant.py
```python
class Constant_Paschen():
    def constant_A_B_Paschen():
        # Paschen....[A,B] A [Ionisation/(Pa m)] B [V/(Pa m)]
        constant_A_B_Paschen = {
            "Air": [11, 274],
            "N2": [9.0, 257],
            "H2": [3.8, 104],
            "He": [2.3, 26],
            "Ne": [3.0, 75],
            "Ar": [11, 135],
            "Kr": [13, 180],
            "Xe": [20, 263],
        }
        return constant_A_B_Paschen  
    def gamma():
        gamma = {
        "Air": {
            "Al": 0.035,
            "Cu": 0.025,
            "Ni": 0.036,
            "Pt": 0.017,
            "C": None,
            "W": None,
            "Fe": 0.02,
        },
        "N2": {
            "Al": 0.1,
            "Cu": 0.066,
            "Ni": 0.077,
            "Pt": 0.59,
            "C": None,
            "W": None,
            "Fe": 0.059,
        },
        "H2": {
            "Al": 0.095,
            "Cu": 0.05,
            "Ni": 0.053,
            "Pt": 0.02,
            "C": 0.014,
            "W": None,
            "Fe": 0.061,
        },
        "He": {
            "Al": 0.021,
            "Cu": None,
            "Ni": 0.015,
            "Pt": 0.01,
            "C": None,
            "W": None,
            "Fe": 0.015,
        },
        "Ne": {
            "Al": 0.053,
            "Cu": 0.02,
            "Ni": 0.031,
            "Pt": 0.023,
            "C": None,
            "W": 0.045,
            "Fe": 0.022,
        },
        "Ar": {
            "Al": 0.12,
            "Cu": 0.058,
            "Ni": 0.058,
            "Pt": 0.058,
            "C": None,
            "W": None,
            "Fe": 0.058,
        },
        "Kr": {
            "Al": None,
            "Cu": None,
            "Ni": None,
            "Pt": None,
            "C": None,
            "W": None,
            "Fe": None,
        },
        "Xe": {
            "Al": None,
            "Cu": None,
            "Ni": None,
            "Pt": None,
            "C": None,
            "W": None,
            "Fe": None,
        },
    }
        return gamma
```

G-D-Zhou commented:
Paschen_curve.py
```python
import math
import matplotlib.pyplot as plt
import numpy as np
from constant import Constant_Paschen

particle = ["Ar", "N2", "H2", "O2", "CH4"]
material = ["Al", "Cu", "Ni", "Pt", "C", "W", "Fe"]


constant_A_B_Paschen = Constant_Paschen.constant_A_B_Paschen()
gamma = Constant_Paschen.gamma()


def first_Townsend_ionization_coefficent(gas):
    A = constant_A_B_Paschen[gas][0]
    B = constant_A_B_Paschen[gas][1]
    # print(gas,':',A,B)
    return A,B


#A,B=first_Townsend_ionization_coefficent('N2')
# print(A)


def second_Townsend_ionization_coefficent(gas,material):
    gam = gamma[gas]
    SEEC = gam.get(material)
    if SEEC == None:
        SEEC = 0.01         # default value= 0.01 if no gamma value (None) exists
    # print(SEEC)
    return SEEC

def Paschen_law(A,B,pd,SEEC):
    g = np.log(1+1/SEEC)
    # print(g)
    Vb = B*pd/(np.log(A*pd) - np.log(g))
    # print(Vb)
    return(Vb)

def clean_save(data,number,name):
    for j in range(number-1):
        if data[j][1] < 0:
            data[j][0] = None
            data[j][1] = None
        else:
            if data[j][1] < data[j+1][1]:
                data[j][0] = None
                data[j][1] = None
            else:
                break
    np.savetxt(name,data)
    return(data)

def Paschen_curve(pressure_range,d_curve,number,gas,material=None,SEEC_set=[]):
    pd = np.zeros(number)
    Vb = np.zeros(number)
    p = np.linspace(pressure_range[0],pressure_range[1],number)
    d = np.linspace(d_curve[0],d_curve[1],number)
    A,B = first_Townsend_ionization_coefficent(gas)
    data = np.zeros((number,2))
    
    if material is None:
        if SEEC_set == []:
            SEEC = 0.01
        else:
            SEEC = SEEC_set    
    else:
        SEEC = second_Townsend_ionization_coefficent(gas, material)
        print("the SEEC is " + str(SEEC))
    # Paschen_law(A, B, 1, SEEC)
    if SEEC_set == []:
        for i in range(number):
            pd[i] = p[i]*d[i]
            print(pd[i])
            Vb[i] = Paschen_law(A, B, pd[i], SEEC)
            data[i][0] = pd[i]
            data[i][1] = Vb[i]
            print(Vb[i])
            name = 'the Paschen curve of ' + gas +' to ' + material
        data = clean_save(data,number,name)            
        plt.loglog(data[:,0],data[:,1],label=gas)
    else:
        SEEC_size = np.size(SEEC_set)
        print(SEEC_size)
        for s in range(int(SEEC_size)):
            for i in range(number):
                pd[i] = p[i]*d[i]
                #print(pd[i])
                SEEC = SEEC_set[s]
                Vb[i] = Paschen_law(A, B, pd[i], SEEC)
                data[i][0] = pd[i]
                data[i][1] = Vb[i]
                #print(Vb[i])
            name = 'the Paschen curve of ' + gas + ' set SEEC'+ '_' + str(SEEC_set[s])
            data = clean_save(data,number,name)
            plt.loglog(data[:,0],data[:,1],label=str(SEEC_set[s]))
    return data


                
pressure_range = np.array([0.08,40])
d_curve = np.array([1,8])

# for m in particle:
#     data = Paschen_curve(pressure_range, d_curve, 10000,m,'Pt')
#     plt.loglog(data[:,0],data[:,1],label = m,)

SEEC_linspace = np.linspace(0.01,1,10)


data = Paschen_curve(pressure_range, d_curve, 1000,'H2','Pt',)
plt.loglog(data[:,0],data[:,1])


plt.xlabel("Pa*m ")  # Add  x-label to the axes.
plt.ylabel(" Breakdown voltage V")  # Add ay-label to the axes.
plt.title("Paschen curve")  # Add a title to the axes.
plt.legend()  # Add legend
plt.show()
```

HollensteinCHH commented:
Dear G.Zhou

Thanks for you very very useful comments and changes to the Paschen contribution. I will go through them and incorporate them into the present pull request. I see that I have still to learn much on Python programming!
Unfortunately I could not work much on the Paschen and Saha contribution since I spent more time with the dentist and I addition I had some medical urgency in the family (is ok know). 
Concerning Saha I am introducing a part on the calculation of the composition of a thermal hydrogen plasma including dissociation and ionisation and I am still looking for the various partition function using for furthering cases (Ar, N2...). If I am a little bit further advanced I will send you a draft of this contribution.

Nice weekend 

Christoph

G-D-Zhou commented:
> Dear G.Zhou
> 
> Thanks for you very very useful comments and changes to the Paschen contribution. I will go through them and incorporate them into the present pull request. I see that I have still to learn much on Python programming!
> Unfortunately I could not work much on the Paschen and Saha contribution since I spent more time with the dentist and I addition I had some medical urgency in the family (is ok know).
> Concerning Saha I am introducing a part on the calculation of the composition of a thermal hydrogen plasma including dissociation and ionisation and I am still looking for the various partition function using for furthering cases (Ar, N2...). If I am a little bit further advanced I will send you a draft of this contribution.
> 
> Nice weekend
> 
> Christoph

Dear Christophe
I am very sorry to hear that there has some medical urgency in your family. I wish this situation disappear as soon as possible, and I hope the world will return to normal.

I look forward to your work in Saha. I regret not being able to contribute in more ways. I'm an amateur studying plasma physics and numerical simulation. Although I have purchased some books on Finite-Difference Time-Domain (FDTD) and other related books, it is difficult for me to spare time to study at the moment, because I need to concentrate on completing my research work to obtain a PhD. This is very critical to me.

Hope to keep in touch!
Sincerely,
G. Zhou

HollensteinCHH commented:
Hi Nick 

Sent you by mail a short description for a Saha contribution!
